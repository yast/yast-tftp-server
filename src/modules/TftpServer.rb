# encoding: utf-8

# File:	modules/TftpServer.ycp
# Package:	Configuration of TftpServer
# Summary:	Data for configuration of TftpServer, input and output functions.
# Authors:	Martin Vidner <mvidner@suse.cz>
#
# $Id$
#
# Representation of the configuration of TftpServer.
# Input and output routines.
require "yast"
require "yast2/systemd/service"
require "yast2/systemd/socket"

require "shellwords"

require "yast2/target_file" # allow CFA to work on change scr
require "cfa/tftp_sysconfig"
require "y2firewall/firewalld"


module Yast
  class TftpServerClass < Module

    SOCKET_NAME = "tftp"
    PACKAGE_NAME = "tftp"

    include Yast::Logger

    def main
      textdomain "tftp-server"

      Yast.import "Package"
      Yast.import "Progress"
      Yast.import "Report"
      Yast.import "Summary"

      # Any settings modified?
      # As we have only a single dialog which handles it by itself,
      # it is used only by autoinst cloning.
      @modified = false

      # Required packages for operation
      @required_packages = [PACKAGE_NAME]

      # if socket start tftp
      @start = false

      # sysconfig model
      @sysconfig = ::CFA::TftpSysconfig.new
      # sysconfig values we are interested in. Allow to change it with UI.
      # TODO when doing bigger changes use sysconfig model everywhere
      @directory = "/srv/tftpboot" # default value

      # Detect who is serving tftp:
      # Systemd may be running, it is the default. But it is ok unless it is
      # serving tftp. So we detect who is serving tftp and warn if it is
      # not systemd socket/service or in.tftpd.
      # If nonempty, the user is notified and the module gives up.
      @foreign_servers = ""
    end

    # firewall instance
    def firewall
      @firewall ||= Y2Firewall::Firewalld.instance
    end

    # systemd socket for tftp
    def socket
      @socket ||= Yast2::Systemd::Socket.find!(SOCKET_NAME)
    end

    # Returns true if the settings were modified
    # @return settings were modified
    def GetModified
      @modified
    end

    # Function sets an internal variable indicating that any
    # settings were modified to "true".
    # Used for autoinst cloning.
    def SetModified
      @modified = true

      nil
    end

    # Read all tftp-server settings
    # @return true on success
    def Read
      # if package is not installed, then read is easy
      return true unless Package.Installed(PACKAGE_NAME)

      # foreign_servers:
      # get command names via lsof, filter out systemd and in.tftpd
      out = Convert.to_map(
        SCR.Execute(path(".target.bash_output"), "/usr/bin/lsof -i :tftp -Fc")
      )
      lines = Ops.get_string(out, "stdout", "").lines
      # the command is field 'c'
      lines = lines.select { |l| l.start_with?("c") }
      # strip the c
      lines.map! { |l| l[1..-1].strip }
      # filter out our servers
      lines.reject! { |l| ["in.tftpd", "systemd"].include?(l) }
      @foreign_servers = lines.join(", ")

      @sysconfig.load
      @directory = @sysconfig.directory

      # force find of socket. It can happen if we need to install package first
      @start = socket.enabled?

      # TODO only when we have our own Progress
      #boolean progress_orig = Progress::set (false);
      firewall.read
      #Progress::set (progress_orig);

      true
    end

    # Return error string to be used in WriteOnly (for autoinst)
    # or before the edit dialog.
    # @return error string
    def ForeignServersError
      # error popup
      # %1 is a command name (or a comma (, ) separated list of them)
      Builtins.sformat(
        _(
          "This module can only use systemd socket to set up TFTP.\n" +
            "However, another program is serving TFTP: %1.\n" +
            "Exiting.\n"
        ),
        @foreign_servers
      )
    end

    # Write all tftp-server settings
    # without actually (re)starting the service
    # @return true on success
    def WriteOnly
      Builtins.y2milestone("Writing")

      # give up if tftp is served by someone else
      if @foreign_servers != ""
        Report.Error(ForeignServersError())
        return false
      end

      @sysconfig.directory = @directory
      @sysconfig.save


      # image dir: if does not exist, create with root:root rwxr-xr-x
      SCR.Execute(path(".target.mkdir"), @directory)
      # and then switch to user which is used for tftp service
      SCR.Execute(path(".target.bash_output"), "/usr/bin/chown #{@sysconfig.user}: #{Shellwords.escape(@directory)}")

      # enable and (re)start systemd socket
      if @start
        socket.enable
        socket.start
      else
        socket.disable
        socket.stop
      end

      # TODO only when we have our own Progress
      #boolean progress_orig = Progress::set (false);
      firewall.write_only
      #Progress::set (progress_orig);

      true
    end

    # Write all tftp-server settings
    # @return true on success
    def Write
      return false if !WriteOnly()

      # in.tftpd will linger around for 15 minutes waiting for a new connection
      # so we must kill it otherwise it will be using the old parameters
      Yast2::Systemd::Service.find!("tftp").stop

      # TODO only when we have our own Progress
      #boolean progress_orig = Progress::set (false);
      firewall.reload
      #Progress::set (progress_orig);

      true
    end

    # Set module data, without validity checking
    # @param [Hash] settings may be empty for reset
    # @return [void]
    def Set(settings)
      settings = deep_copy(settings)
      @start = Ops.get_boolean(settings, "start_tftpd", false)
      @directory = Ops.get_string(settings, "tftp_directory", "/srv/tftpboot")

      nil
    end


    # Get all tftp-server settings from the first parameter
    # (For use by autoinstallation.)
    # @param [Hash] settings The YCP structure to be imported.
    # @return [Boolean] True on success
    def Import(settings)
      settings = deep_copy(settings)
      if settings != {}
        s = "start_tftpd"
        d = "tftp_directory"
        if !Builtins.haskey(settings, s)
          Builtins.y2error("Missing at Import: '%1'.", s)
          return false
        end
        if Ops.get_boolean(settings, s, false) && !Builtins.haskey(settings, d)
          Builtins.y2error("Missing at Import: '%1'.", d)
          return false
        end
      end
      Set(settings)
      true
    end

    # Dump the tftp-server settings to a single map
    # (For use by autoinstallation.)
    # @return [Hash] Dumped settings (later acceptable by Import ())
    def Export
      settings = { "start_tftpd" => @start }
      Ops.set(settings, "tftp_directory", @directory) if @start
      deep_copy(settings)
    end


    # Mergeing config to existing system configuration. It is useful for delayed write.
    # So if package will be installed later this method re-apply changes on top of newly parsed
    # file.
    def merge_to_system
      config = Export()
      Read()
      Import(config)
    end

    # @return Html formatted configuration summary
    def Summary
      summary = ""
      nc = Summary.NotConfigured

      # summary header
      summary = Summary.AddHeader(summary, _("TFTP Server Enabled:"))
      # summary item: an option is turned on
      summary = Summary.AddLine(summary, @start ? _("Yes") : nc)

      # summary header
      summary = Summary.AddHeader(summary, _("Boot Image Directory:"))
      summary = Summary.AddLine(summary, @directory != "" ? @directory : nc)

      summary
    end

    # Return needed packages and packages to be removed
    # during autoinstallation.
    # @return [Hash] of lists.
    #
    #

    def AutoPackages
      install_pkgs = deep_copy(@required_packages)
      remove_pkgs = []
      { "install" => install_pkgs, "remove" => remove_pkgs }
    end

    publish :function => :GetModified, :type => "boolean ()"
    publish :function => :SetModified, :type => "void ()"
    publish :variable => :required_packages, :type => "list <string>"
    publish :variable => :start, :type => "boolean"
    publish :variable => :directory, :type => "string"
    publish :variable => :foreign_servers, :type => "string"
    publish :function => :Read, :type => "boolean ()"
    publish :function => :ForeignServersError, :type => "string ()"
    publish :function => :WriteOnly, :type => "boolean ()"
    publish :function => :Write, :type => "boolean ()"
    publish :function => :Set, :type => "void (map)"
    publish :function => :Import, :type => "boolean (map)"
    publish :function => :Export, :type => "map ()"
    publish :function => :Summary, :type => "string ()"
    publish :function => :AutoPackages, :type => "map ()"
  end

  TftpServer = TftpServerClass.new
  TftpServer.main
end
