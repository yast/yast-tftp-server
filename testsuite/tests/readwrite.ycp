/**
 * Module:  TFTP server configuration
 * Summary: Testsuite
 * Authors: Martin Vidner <mvidner@suse.cz>
 *
 * $Id$
 */
{
    // testedfiles: TftpServer.ycp Service.ycp Testsuite.ycp

    include "testsuite.ycp";

    map READ = $[
	"etc": $[
	    "xinetd_d": $[
		"tftp": $[
		    "section": $[
			"tftp": $[],
			],
		    "value": $[
			"tftp": $[
			    "disable": nil,
			    "server_args": "-s -v /srv/boot"
			    ],
			],
		    ],
		],
	    ],
	    "target" : $[
		"size" : 0,
	    ],
	];

    map WRITE = $[
	];

    map EXECUTE = $[
	"target": $[
	    // ok if used both for `lsof` and `rcfoo start`
	    "bash_output": $[
		"exit":0,
		"stderr":"",
		"stdout":"p3316\ncxinetd\n"
		],
/*
	    "remove": true, // /etc/yp.conf.sv
*/
	    ],
	];

    TESTSUITE_INIT ([READ, WRITE, EXECUTE], nil);

    import "TftpServer";

    TEST (``(TftpServer::Read ()), [READ, WRITE, EXECUTE], nil);
    TEST (``(TftpServer::Write ()), [READ, WRITE, EXECUTE], nil);

    //READ["etc","xinetd_d","tftp","value","tftp","server_args"] = ...
}
