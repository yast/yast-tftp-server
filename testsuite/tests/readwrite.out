Execute	.target.bash_output "/usr/bin/lsof -i :tftp -Fc" $["exit":0, "stderr":"", "stdout":"p3316\ncxinetd\n"]
Execute	.target.bash "/bin/systemctl is-enabled xinetd.service" 0
Dir	.etc.xinetd_d.tftp.section: ["tftp"]
Read	.etc.xinetd_d.tftp.value.tftp.disable nil
Read	.etc.xinetd_d.tftp.value.tftp.server_args "-s -v /srv/boot"
Execute	.target.bash "/bin/systemctl is-enabled SuSEfirewall2.service" 0
Return	true
Write	.etc.xinetd_d.tftp.value.tftp."{" "" true
Write	.etc.xinetd_d.tftp.value_type.tftp."{" 1 true
Write	.etc.xinetd_d.tftp.value.tftp.disable "no" true
Write	.etc.xinetd_d.tftp.value.tftp.socket_type "dgram" true
Write	.etc.xinetd_d.tftp.value.tftp.protocol "udp" true
Write	.etc.xinetd_d.tftp.value.tftp.wait "yes" true
Write	.etc.xinetd_d.tftp.value.tftp.user "root" true
Write	.etc.xinetd_d.tftp.value.tftp.server "/usr/sbin/in.tftpd" true
Write	.etc.xinetd_d.tftp.value.tftp.server_args "-v -s /srv/boot" true
Write	.etc.xinetd_d.tftp nil true
Execute	.target.mkdir "/srv/boot" 0
Execute	.target.bash "/bin/systemctl is-enabled xinetd.service" 0
Execute	.target.bash "/usr/bin/killall in.tftpd" 0
Execute	.target.bash_output "/bin/systemctl restart xinetd.service" $["TERM":"raw"] $["exit":0, "stderr":"", "stdout":"p3316\ncxinetd\n"]
Return	true
